client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
 //
-// Since a single rejection causes an error to be returned, SignIdentities
-// should currently only be used when requesting signed identities for a single
-// allocation.
-func (m *WIDMgr) SignIdentities(minIndex uint64, req []*structs.WorkloadIdentityRequest) ([]*structs.SignedWorkloadIdentity, error) {
-	args := structs.AllocIdentitiesRequest{
-		Identities: req,
-		QueryOptions: structs.QueryOptions{
-			Region:        m.region,
-			MinQueryIndex: minIndex - 1,
-			AllowStale:    true,
-			AuthToken:     m.nodeSecret,
-		},
+// If an error is returned the identities could not be fetched and the renewal
+// goroutine was not started.
+func (m *WIDMgr) Run() error {
+	if len(m.widSpecs) == 0 {
+		m.logger.Debug("no workload identities to retrieve or renew")
+		return nil
 	}
-	reply := structs.AllocIdentitiesResponse{}
-	if err := m.rpc.RPC("Alloc.SignIdentities", &args, &reply); err != nil {
-		return nil, err
+
+	m.logger.Debug("retrieving and renewing workload identities", "num_identities", len(m.widSpecs))
+
+	if err := m.getIdentities(); err != nil {
+		return fmt.Errorf("failed to fetch signed identities: %w", err)
 	}
 
-	if n := len(reply.Rejections); n == 1 {
-		return nil, fmt.Errorf("%d/%d signing request was rejected", n, len(req))
-	} else if n > 1 {
-		return nil, fmt.Errorf("%d/%d signing requests were rejected", n, len(req))
+	go m.renew()
+
+	return nil
+}
+
+// Get retrieves the latest signed identity or returns an error. It must be
+// called after Run and does not block.
+//
+// For retrieving tokens which might be renewed callers should use Watch
+// instead to avoid missing new tokens retrieved by Run between Get and Watch
+// calls.
+func (m *WIDMgr) Get(id cstructs.TaskIdentity) (*structs.SignedWorkloadIdentity, error) {
+	token := m.get(id)
+	if token == nil {
+		// This is an error as every identity should have a token by the time Get
+		// is called.
+		return nil, fmt.Errorf("unable to find token for task %q and identity %q", id.TaskName, id.IdentityName)
 	}
 
-	if len(reply.SignedIdentities) == 0 {
-		return nil, fmt.Errorf("empty signed identity response")
+	return token, nil
+}
+
+func (m *WIDMgr) get(id cstructs.TaskIdentity) *structs.SignedWorkloadIdentity {
+	m.lastTokenLock.Lock()
+	defer m.lastTokenLock.Unlock()
+
+	return m.lastToken[id]
+}
+
+// Watch sends new signed identities until it is closed due to shutdown. Must
+// be called after Run.
+//
+// The caller must call the returned func to stop watching.
+func (m *WIDMgr) Watch(id cstructs.TaskIdentity) (<-chan *structs.SignedWorkloadIdentity, func()) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	// If Shutdown has been called return a closed chan
+	if m.stopCtx.Err() != nil {
+		c := make(chan *structs.SignedWorkloadIdentity)
+		close(c)
+		return c, func() {}
+	}
+
+	//TODO I think we only want 1 watcher per id, so let's just close on double watches?
+	if existing, ok := m.watchers[id]; ok {
+		close(existing)
+	}
```

Comment:
To add a new data point here, it's _very_ easy to accidentally watch an identity twice. 

I'm updating the `vault` hook and it needs to access the signed identity to derive a Vault token, but the `identity` hook also needs this information to set env/file.

My prototype code looks like this:
https://github.com/hashicorp/nomad/blob/829bf43cb482b91cedf40c77675a280bda17b000/client/allocrunner/taskrunner/vault_hook.go#L364-L381

This was causing the `identity` hook watcher to silently close and the renewed identity not be updated to disk (which is what I was checking). It took me a _long_ time to realize the second `Watch()` was closing it.

Arguably I should be using `Get()` in the `vault`  hook since I don't actually need to keep track of updates, but it was very easy to make this mistake and very hard to debug it.